src/test/tools/sparsefile/sparsefile.c

```diff
@@ -99,123 +91,187 @@ print_file_size(const wchar_t *filename)
 	BOOL ret = GetVolumeInformationByHandleW(fh, NULL, 0, NULL, NULL,
 			&flags, NULL, 0);
 	if (ret == FALSE) {
-		if (Opt_verbose || Opt_sparse)
-			out_err(L"GetVolumeInformationByHandle");
-	} else if ((flags & FILE_SUPPORTS_SPARSE_FILES) == 0) {
-		if (Opt_verbose || Opt_sparse)
-			out_err(L"Volume does not support sparse files.");
-		if (Opt_sparse)
-			goto err;
-	}
-
-	/* mark file as sparse */
-	if (flags & FILE_SUPPORTS_SPARSE_FILES) {
-		DWORD nbytes;
-		ret = DeviceIoControl(fh, FSCTL_SET_SPARSE, NULL, 0, NULL,
-				0, &nbytes, NULL);
-		if (ret == FALSE) {
-			if (Opt_verbose || Opt_sparse)
-				out_err(L"DeviceIoControl");
-			if (Opt_sparse)
-				goto err;
-		}
+		out_err(L"GetVolumeInformationByHandle");
+		return 1;
 	}
 
-	/* set file length */
-	LARGE_INTEGER llen;
-	llen.QuadPart = len;
+	if ((flags & FILE_SUPPORTS_SPARSE_FILES) == 0) {
+		return force ? 1 : 0;
+	}
 
-	ret = SetFilePointerEx(fh, llen, NULL, FILE_BEGIN);
+	DWORD nbytes;
+	ret = DeviceIoControl(fh, FSCTL_SET_SPARSE, NULL, 0, NULL,
+			0, &nbytes, NULL);
 	if (ret == FALSE) {
-		out_err(L"SetFilePointerEx");
-		goto err;
+		out_err(L"DeviceIoControl(FSCTL_SET_SPARSE)");
+		return 1;
 	}
 
-	ret = SetEndOfFile(fh);
+	return 0;
+}
+
+/*
+ * set_compressed_file -- creates sparse file of given size
+ */
+static int
+set_compressed_file(HANDLE fh, int force)
+{
+	/* check if sparse files are supported */
+	DWORD flags = 0;
+	BOOL ret = GetVolumeInformationByHandleW(fh, NULL, 0, NULL, NULL,
+		&flags, NULL, 0);
 	if (ret == FALSE) {
-		out_err(L"SetEndOfFile");
-		goto err;
+		out_err(L"GetVolumeInformationByHandle");
+		return 1;
 	}
 
-	CloseHandle(fh);
-	return 0;
+	if ((flags & FILE_FILE_COMPRESSION) == 0) {
+		return force ? 1 : 0;
+	}
 
-err:
-	CloseHandle(fh);
-	DeleteFileW(filename);
-	return -1;
+	DWORD nbytes;
+	USHORT n = 1; /* magic undocumented value */
+	ret = DeviceIoControl(fh, FSCTL_SET_COMPRESSION, &n, sizeof(n), NULL, 0,
+		&nbytes, NULL);
+	if (ret == FALSE) {
+		out_err(L"DeviceIoControl(FSCTL_SET_COMPRESSION)");
+		return 1;
+	}
+	return 0;
 }
 
 int
-wmain(int argc, const wchar_t *argv[])
+main(int argc, char *argv[])
 {
 	util_suppress_errmsg();
+	wchar_t **wargv = CommandLineToArgvW(GetCommandLineW(), &argc);
+	for (int i = 0; i < argc; i++) {
+		argv[i] = util_toUTF8(wargv[i]);
+		if (argv[i] == NULL) {
+			for (i--; i >= 0; i--)
+				free(argv[i]);
+			out_err(L"Error during arguments conversion\n");
+			return 1;
+		}
+	}
+
 	if (argc < 2) {
-		fwprintf(stderr, L"Usage: %s filename len\n", argv[0]);
+		fprintf(stderr, "Usage: %s filename len\n", argv[0]);
 		exit(1);
 	}
 
 	int i = 1;
-	while (i < argc && argv[i][0] == '-') {
-		switch (argv[i][1]) {
+	int opt;
+	DWORD create_mode = OPEN_ALWAYS;
+	int compress = 0;
+	int sparse = 0;
+	int check = 0;
+	int force = 0;
+	long long len = 0;
+	while ((opt = getopt(argc, argv, "vnscfl:")) != -1) {
+		switch (opt) {
+			case 'p':
+				check = 1;
 			case 'v':
 				Opt_verbose = 1;
```

Comment:
use the same naming scheme consistently.

Reply:
Done